library/std/src/thread/mod.rs

```diff
@@ -193,21 +193,30 @@ pub use scoped::{scope, Scope, ScopedJoinHandle};
 #[stable(feature = "rust1", since = "1.0.0")]
 pub use self::local::{AccessError, LocalKey};
 
-// The types used by the thread_local! macro to access TLS keys. Note that there
-// are two types, the "OS" type and the "fast" type. The OS thread local key
-// type is accessed via platform-specific API calls and is slow, while the fast
+// Select the type used by the thread_local! macro to access TLS keys. There
+// are three types: "static", "fast", "OS". The "OS" thread local key
+// type is accessed via platform-specific API calls and is slow, while the "fast"
 // key type is accessed via code generated via LLVM, where TLS keys are set up
-// by the elf linker. Note that the OS TLS type is always available: on macOS
-// the standard library is compiled with support for older platform versions
-// where fast TLS was not available; end-user code is compiled with fast TLS
-// where available, but both are needed.
+// by the elf linker. "static" is for single-threaded platforms where a global
+// static is sufficient.
 
 #[unstable(feature = "libstd_thread_internals", issue = "none")]
 #[cfg(target_thread_local)]
+#[cfg(not(test))]
 #[doc(hidden)]
 pub use self::local::fast::Key as __FastLocalKeyInner;
+#[unstable(feature = "libstd_thread_internals", issue = "none")]
+#[cfg(target_thread_local)]
+#[cfg(test)] // when building for tests, use real std's key
+pub use realstd::thread::__FastLocalKeyInner;
+#[unstable(feature = "libstd_thread_internals", issue = "none")]
+#[cfg(target_thread_local)]
+#[cfg(test)]
+pub use self::local::fast::Key as __FastLocalKeyInnerUnused; // we import this anyway to silence 'unused' warnings
+
 #[unstable(feature = "libstd_thread_internals", issue = "none")]
 #[doc(hidden)]
+#[cfg(not(target_thread_local))]
 pub use self::local::os::Key as __OsLocalKeyInner;
```

Comment:
It turns out that only one of these three key types is ever actually used now. So I considered refactoring this to avoid giving them different names, instead making `__LocalKeyInner` always be *the* key type for the current configuration. Do you think that would be a good idea?

The only issue I can see is that the types have to be used in slightly different ways: `__FastLocalKeyInner` wants a `#[thread_local] static`, the others a regular `static`. Using the same name for both usages slightly increases the risk that this would be mixed up.

Also I wasn't sure whether `all(target_family = "wasm", not(target_feature = "atomics"))` implied `not(target_thread_local)`, if that is the case some of these `cfg` could be simplified and we could potentially at least merge `__OsLocalKeyInner` and `__StaticLocalKeyInner` into one type (i.e., move the logic for which implementation to use for that into the `local` module, rather than having it spread between that module and this file).

Reply:
>  I wasn't sure whether all(target_family = "wasm", not(target_feature = "atomics")) implied not(target_thread_local)

Turns out the answer is no: https://github.com/rust-lang/rust/issues/84224

Reply:
Yeah, basically have a single name for "the TLS key type we are using".

Not sure if it's worth it though.